enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/tx/TransactionLogCatchUpWriter.java

```diff
@@ -40,41 +40,43 @@
 import org.neo4j.logging.LogProvider;
 
 import static org.neo4j.kernel.impl.store.MetaDataStore.Position.LAST_TRANSACTION_ID;
+import static org.neo4j.kernel.impl.transaction.log.entry.LogHeader.LOG_HEADER_SIZE;
 
 public class TransactionLogCatchUpWriter implements TxPullResponseListener, AutoCloseable
 {
     private final Lifespan lifespan = new Lifespan();
-    private final File neoStoreFile;    ;
+    private final File neoStoreFile;
     private final PageCache pageCache;
     private final Log log;
     private final TransactionLogWriter writer;
+    private final PhysicalLogFiles logFiles;
+    private final File storeDir;
 
-    private long txId = -1;
+    private long lastTxId = -1;
 
-    TransactionLogCatchUpWriter( File storeDir, FileSystemAbstraction fs, PageCache pageCache, LogProvider logProvider )
+    TransactionLogCatchUpWriter( File storeDir, FileSystemAbstraction fs, PageCache pageCache, LogProvider logProvider, long lastCommittedTxId )
             throws IOException
     {
         this.neoStoreFile = new File( storeDir, MetaDataStore.DEFAULT_NAME );
         this.pageCache = pageCache;
         this.log = logProvider.getLog( getClass() );
-
-        PhysicalLogFiles logFiles = new PhysicalLogFiles( storeDir, fs );
+        this.logFiles = new PhysicalLogFiles( storeDir, fs );
         ReadOnlyLogVersionRepository logVersionRepository = new ReadOnlyLogVersionRepository( pageCache, storeDir );
-        ReadOnlyTransactionIdStore readOnlyTransactionIdStore = new ReadOnlyTransactionIdStore( pageCache, storeDir );
         LogFile logFile = lifespan.add( new PhysicalLogFile( fs, logFiles, Long.MAX_VALUE /*don't rotate*/,
-                () -> readOnlyTransactionIdStore.getLastCommittedTransactionId() - 1, logVersionRepository,
+                () -> lastCommittedTxId - 1, logVersionRepository,
                 new Monitors().newMonitor( PhysicalLogFile.Monitor.class ), new LogHeaderCache( 10 ) ) );
         this.writer = new TransactionLogWriter( new LogEntryWriter( logFile.getWriter() ) );
+        this.storeDir = storeDir;
     }
 
     @Override
-    public void onTxReceived( TxPullResponse txPullResponse )
+    public synchronized void onTxReceived( TxPullResponse txPullResponse )
```

Comment:
onTxReceived is a callback for netty while close is for the main thread, they need synchronization due to JMM